Add automotive IdmController

jadecastro opened the pull request:
+@liangfok for feature review, please.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5618)
<!-- Reviewable:end -->

liangfok commented:
Checkpoint.

---

Reviewed 9 of 12 files at r1.
Review status: 9 of 12 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[drake/automotive/idm_controller.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4KBQIIknAqR4qKABq:-Kg4KBQIIknAqR4qKABr:b-osj1cx) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L22)):*
> ```objc
> namespace automotive {
> 
> /// A controller that implements the IDM (Intelligent Driver Model) planner,
> ```

This should start: "An IdmController implements..."

---

*[drake/automotive/idm_controller.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4KVT-V23yv9DvIx4M:-Kg4KVT-V23yv9DvIx4N:bb2mi65) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L50)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(IdmController)
> 
>   /// Constructor. @p road the pre-defined RoadGeometry.
> ```

Missing "is" after "road" or the second sentence should be moved into its own line and modified to be:

```
/// @param road the pre-defined RoadGeometry.
```

---

*[drake/automotive/idm_controller.h, line 71 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Kt5aZMld2dNipBsj:-Kg4Kt5aZMld2dNipBsk:b-z46ecs) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L71)):*
> ```objc
> 
>  private:
>   /// Retrieves the current RoadPosition for a given road and PoseVector.
> ```

This is a bit confusing to me: The word "given" implies to me that it's a parameter. Also, the input isn't a PoseVector.

---

*[drake/automotive/idm_controller.h, line 87 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4LK8LAMehshraUf1G:-Kg4LK8LAMehshraUf1H:b3yfmko) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L87)):*
> ```objc
>   const std::unique_ptr<maliput::api::RoadGeometry> road_{};
> 
>   int ego_pose_index_{};
> ```

BTW, consider adding a brief comment indicating what these are indices into.

---

*[drake/automotive/idm_planner_parameters.named_vector, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Ja1tuPfLfNLlQD7f:-Kg4Ja1tuPfLfNLlQD7g:byuzfl8) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L26)):*
> ```
> }
> element {
>      name: "bloat_diameter"
> ```

BTW, should this eventually make use of the vehicle's geometric collision model? If so, add a TODO?

---

*[drake/automotive/idm_planner_parameters.named_vector, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Jlfqg9jQzAKtz_a7:-Kg4Jlfqg9jQzAKtz_a8:beyu8s1) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L31)):*
> ```
> element {
>      name: "distance_lower_limit"
>      doc: "lower saturation bound on net distance"
> ```

Does this mean any distance less than this limit is considered a collision? How does this differ from `bloat_diameter`?

---

*[drake/automotive/idm_planner_parameters.named_vector, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4JTmBTEPdNcUNaFse:-Kg4JTmBTEPdNcUNaFsf:bs3q2p4) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L32)):*
> ```
>      name: "distance_lower_limit"
>      doc: "lower saturation bound on net distance"
> }
> ```

BTW, missing newline.

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4HBA8SVqlm_jQXkPS:-Kg4HBA8SVqlm_jQXkPT:b-89q3ka) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L38)):*
> ```objc
> /// given @p traffic_lane to an ego car as if the ego car were traveling in @p
> /// traffic_lane at its current `s`-position.  The ego car's pose @ego_pose and
> /// the odometies of the traffic cars (@p traffic_odom) are assumed to exist on
> ```

typo: odometries

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4HgNPHNAN54CLrOXh:-Kg4HgNPHNAN54CLrOXi:bpf9ri8) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L38)):*
> ```objc
> /// given @p traffic_lane to an ego car as if the ego car were traveling in @p
> /// traffic_lane at its current `s`-position.  The ego car's pose @ego_pose and
> /// the odometies of the traffic cars (@p traffic_odom) are assumed to exist on
> ```

The `@p traffic_odom` syntax indicates that `traffic_odom` is a parameter. Yet it is not. 

---

*[drake/automotive/pose_selector.h, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4I1RWnZZTN6hIyMwf:-Kg4I1RWnZZTN6hIyMwg:binegdx) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L39)):*
> ```objc
> /// traffic_lane at its current `s`-position.  The ego car's pose @ego_pose and
> /// the odometies of the traffic cars (@p traffic_odom) are assumed to exist on
> /// the same @p road.  If @p traffic_lane is `nullptr`, the the ego car's
> ```

Duplicate "the".

---

*[drake/automotive/pose_selector.h, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4IUKj29yGzB8Ed8eQ:-Kg4IUKj29yGzB8Ed8eR:byn8fk1) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L40)):*
> ```objc
> /// the odometies of the traffic cars (@p traffic_odom) are assumed to exist on
> /// the same @p road.  If @p traffic_lane is `nullptr`, the the ego car's
> /// current lane is used.  If no cars are seen within @p traffic_lane, car
> ```

BTW, consider mentioning how the ego car's current lane is derived.

---

*[drake/automotive/pose_selector.h, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4IfXAos2YiW_q07c-:-Kg4IfXB-1vib4Dw8NdR:blt5xws) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L40)):*
> ```objc
> /// the odometies of the traffic cars (@p traffic_odom) are assumed to exist on
> /// the same @p road.  If @p traffic_lane is `nullptr`, the the ego car's
> /// current lane is used.  If no cars are seen within @p traffic_lane, car
> ```

BTW, consider adding "leading and trailing" before "car".

---

*[drake/automotive/pose_selector.h, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4InYsm703jUeVs3DE:-Kg4InYsm703jUeVs3DF:bgzg38k) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L43)):*
> ```objc
> /// `s`-positions are taken to be at infinite distances away from the ego car.
> ///
> /// The return values are a pair of leading/trailing RoadPositions. Note that
> ```

This sentence should be updated now that the return value is a `std::pair<RoadOdometry<double>, RoadOdometry<double>>`.

---

*[drake/automotive/pose_selector.h, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4JA97dt0EdiR5IcSI:-Kg4JA97dt0EdiR5IcSJ:bhaxrsm) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L64)):*
> ```objc
> 
> /// Same as FindClosestPair() except that: (1) it only considers the ego car's
> /// lane and (2) it returns a single the RoadPosition of the leading vehicle.
> ```

"the RoadPosition" should be updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-Kg4LTkpvN9j-WPXdKP2:by96gdd)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 5 of 12 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/automotive/idm_controller.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4KBQIIknAqR4qKABq:-Kg4NMXeNYG05CypTLgY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This should start: "An IdmController implements..."
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4KVT-V23yv9DvIx4M:-Kg4NRgMhLNFl01NbpoI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L50)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Missing "is" after "road" or the second sentence should be moved into its own line and modified to be:

```
/// @param road the pre-defined RoadGeometry.
```
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.h, line 71 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Kt5aZMld2dNipBsj:-Kg4NkwiRPFF3n4XRDQd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L71)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This is a bit confusing to me: The word "given" implies to me that it's a parameter. Also, the input isn't a PoseVector.
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.h, line 87 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4LK8LAMehshraUf1G:-Kg4Nm6O9ovlIHRHvzU-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.h#L87)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a brief comment indicating what these are indices into.
</blockquote></details>

Done.

---

*[drake/automotive/idm_planner_parameters.named_vector, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Ja1tuPfLfNLlQD7f:-Kg4QJ482KU4MhJQS0Z5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should this eventually make use of the vehicle's geometric collision model? If so, add a TODO?
</blockquote></details>

Done.

---

*[drake/automotive/idm_planner_parameters.named_vector, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4Jlfqg9jQzAKtz_a7:-Kg4No1BenA4_-sT3Sab:bfocdqt) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L31)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Does this mean any distance less than this limit is considered a collision? How does this differ from `bloat_diameter`?
</blockquote></details>

It's meant to prevent singular solutions - since the IDM has a singularity at distance = 0. So, this is here to prevent the accelerations from becoming excessively large. I've added to the description to state this.

---

*[drake/automotive/idm_planner_parameters.named_vector, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4JTmBTEPdNcUNaFse:-Kg4NpWzrqBLU0MxDY_H:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_planner_parameters.named_vector#L32)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing newline.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4HBA8SVqlm_jQXkPS:-Kg4OVgwH7Ect5aaF8M5:b-34sgk0) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

typo: odometries
</blockquote></details>

replaced with `pose`.

---

*[drake/automotive/pose_selector.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4HgNPHNAN54CLrOXh:-Kg4OZToU2g82rn2313l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The `@p traffic_odom` syntax indicates that `traffic_odom` is a parameter. Yet it is not. 
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4I1RWnZZTN6hIyMwf:-Kg4OcNdn4fU0F6eI11g:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L39)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Duplicate "the".
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4IfXAos2YiW_q07c-:-Kg4P7sHekTDuqq9UGbu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L40)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding "leading and trailing" before "car".
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4IUKj29yGzB8Ed8eQ:-Kg4Phu7N2USFhdcFdt_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L40)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider mentioning how the ego car's current lane is derived.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4InYsm703jUeVs3DE:-Kg4Ot8URGPgIcGBaff9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L43)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This sentence should be updated now that the return value is a `std::pair<RoadOdometry<double>, RoadOdometry<double>>`.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4JA97dt0EdiR5IcSI:-Kg4OrLx85WzOJ2GLecz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/pose_selector.h#L64)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"the RoadPosition" should be updated.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 4 of 4 files at r2.
Review status: 9 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/pose_selector.h, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4V25wnOcKHluTJAYs:-Kg4V25wnOcKHluTJAYt:b-48bkbe) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L40)):*
> ```objc
> /// the poses of the traffic cars (@p traffic_pose) are assumed to exist on the
> /// same @p road.  If @p traffic_lane is `nullptr`, the ego car's current lane
> /// is used (this is derived from a call to CalcRoadPosition).  If no
> ```

BTW, if you change "CalcRoadPosition" to be "maliput::api::Lane::CalcRoadPosition()", I believe Doxygen will make it a link.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Checkpoint.

---

Reviewed 1 of 12 files at r1.
Review status: 10 of 12 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/automotive/idm_controller.cc, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4d31Hx1waC1hlj2xP:-Kg4d31Hx1waC1hlj2xQ:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L15)):*
> ```cpp
> 
> using maliput::api::RoadGeometry;
> using maliput::api::RoadPosition;
> ```

unused?

---

*[drake/automotive/idm_controller.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4dDK2A5EuJ2UcuOxE:-Kg4dDK2A5EuJ2UcuOxF:b-wawgma) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L27)):*
> ```cpp
> 
> template <typename T>
> IdmController<T>::IdmController(std::unique_ptr<RoadGeometry> road)
> ```

Does it make sense for the IdmController to own the `RoadGeometry`? Should the `RoadGeometry` be owned by the application since numerous simulation subsystems may need access it?

---

*[drake/automotive/idm_controller.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4dlUJMWLskVOmp6rb:-Kg4dlUJMWLskVOmp6rc:b-p83oop) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L40)):*
> ```cpp
>   traffic_index_ = this->DeclareAbstractInputPort().get_index();
>   // Declare the output port.
>   this->DeclareVectorOutputPort(DrivingCommand<T>());
> ```

BTW, should the `IdmController` output a `DrivingCommand` or just a desired acceleration?

---

*[drake/automotive/idm_controller.cc, line 111 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4f3okJEFAKv7Kqu5L:-Kg4f3okJEFAKv7Kqu5M:b6daftx) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L111)):*
> ```cpp
>   const T& s_ego =
>       pose_selector::CalcRoadPosition(*road_, ego_pose.get_isometry()).pos.s;
>   const T& s_dot_ego = ego_velocity.get_velocity().translational().x();
> ```

BTW, have we documented anywhere that this class assumes the vehicle is traveling forward along its positive x axis?

(Ditto regarding the lead car a few lines below.)

---

*[drake/automotive/idm_controller.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4ipA_5Q1w_BQNM2WU:-Kg4ipA_5Q1w_BQNM2WV:b-uhwnpd) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L115)):*
> ```cpp
>   const T& s_dot_lead = lead_car_odom.vel.get_velocity().translational().x();
> 
>   // Saturate the net_distance at distance_lower_bound away from the ego car.
> ```

BTW, consider adding a comment stating that this is to avoid a singularity in IDM's math when the distance equals zero.

---

*[drake/automotive/idm_controller.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4j9-45wJiRpsAZOnz:-Kg4j9-45wJiRpsAZOo-:bpo8en6) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L129)):*
> ```cpp
>       cond(normalized_accel < T(0.), T(0.), normalized_accel));
>   command->set_brake(cond(normalized_accel >= T(0.), T(0.), -normalized_accel));
>   command->set_steering_angle(0.);
> ```

BTW, is the above temporary? If so, should a TODO be added that steering to follow a turning road should eventually be supported?

---

*[drake/automotive/idm_controller.cc, line 136 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4e8mQPQU_mYc9X0_7:-Kg4e8mQPQU_mYc9X0_8:b68v6l) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L136)):*
> ```cpp
>     const {
>   std::vector<std::unique_ptr<systems::BasicVector<T>>> params;
>   params.insert(params.begin() + kIdmParamsIndex,
> ```

BTW, is the inclusion of `params.begin()` here and below just an abundance of caution?

---

*[drake/automotive/idm_controller.cc, line 140 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4eK1QfHJ7DIc7kVfw:-Kg4eK1R4XotBGpj2EZo:b-fmemmj) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L140)):*
> ```cpp
>   params.insert(params.begin() + kCarParamsIndex,
>                 std::make_unique<SimpleCarConfig<T>>());
>   return std::make_unique<systems::Parameters<T>>(std::move(params));
> ```

Since `params` is not a `std::unique_ptr`, do we need the `std::move()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-Kg4jLBGaKK6CqCHy2ir:by96gdd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+@jwnimmer-tri for platform review.
Checkpoint.

---

Reviewed 6 of 12 files at r1, 4 of 4 files at r2.
Review status: 11 of 12 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/automotive/BUILD, line 174 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4reIbh80RSExMV2Ih:-Kg4reIbh80RSExMV2Ii:b-tdf2y5) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/BUILD#L174)):*
> ```Python
>         ":idm_planner",
>         ":simple_car",
>         "pose_selector",
> ```

BTW `":pose_selector"`

---

*[drake/automotive/idm_controller.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4dDK2A5EuJ2UcuOxE:-Kg4vvJAq2suVobm9Yet:b-ruhix0) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L27)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Does it make sense for the IdmController to own the `RoadGeometry`? Should the `RoadGeometry` be owned by the application since numerous simulation subsystems may need access it?
</blockquote></details>

Yes, without reading the context, my guess is that `IdmController` should alias the geometry, and not own it.  There should be a specific reason written down if there's a reason to do it with `unique_ptr` here.

---

*[drake/automotive/idm_controller.cc, line 136 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4e8mQPQU_mYc9X0_7:-Kg4wbQJ4sYlMZGyJF3F:btybyvz) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L136)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is the inclusion of `params.begin()` here and below just an abundance of caution?
</blockquote></details>

I don't understand.  `vector::insert` takes a `vector::iterator`, you can't just pass a bare `int`.

---

*[drake/automotive/idm_controller.h, line 64 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4uedwZ8UeEIfdIA8n:-Kg4uedwZ8UeEIfdIA8o:btnx38z) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/idm_controller.h#L64)):*
> ```objc
>   // System<T> overrides.
>   // The output of this system is an algebraic relation of its inputs.
>   bool has_any_direct_feedthrough() const override { return true; }
> ```

This is the deprecated flavor of the override, so at a minimum should change to use the new spelling.  BTW also the default is `true` anyway, so this method is not required / not helpful.

---

*[drake/automotive/idm_controller.h, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4vMpGfQ1YWrUioZMO:-Kg4vMpGfQ1YWrUioZMP:b-aa4wc1) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/idm_controller.h#L72)):*
> ```objc
> 
>  private:
>   /// Converts @p pose into RoadPosition.
> ```

BTW private methods should use `//` not `///`.

---

*[drake/automotive/pose_selector.cc, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tsXpSdsZPMn0vG_T:-Kg4tsXpSdsZPMn0vG_U:bb14bp6) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.cc#L60)):*
> ```cpp
>         else
>           result_trailing = RoadOdometry<double>(traffic_position,
>                                                  traffic_poses.get_velocity(i));
> ```

Multi-line if-else bodies require curly braces, per cppguide.

---

*[drake/automotive/pose_selector.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tQ8VcEDfL4fYX08q:-Kg4tQ8VcEDfL4fYX08r:b-ow4fdu) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L18)):*
> ```objc
> 
> /// Contains the position of the vehicle with respect to a lane in a road, along
> /// with its body-relative velocity vector.
> ```

BTW `world-relative velocity vector`, yes?  Or just `velocity`, since the meaning of a `FrameVelocity` is well-documented.

---

*[drake/automotive/pose_selector.h, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4sXOkxSPT69s5gn9w:-Kg4sXOkxSPT69s5gn9x:bad2cu8) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L38)):*
> ```objc
> /// given @p traffic_lane to an ego car as if the ego car were traveling in @p
> /// traffic_lane at its current `s`-position.  The ego car's pose @ego_pose and
> /// the poses of the traffic cars (@p traffic_pose) are assumed to exist on the
> ```

BTW `traffic_poses`

---

*[drake/automotive/test/pose_selector_test.cc, line 140 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4u9aurAMEbcxIrN3-:-Kg4u9avVp8P4Lza0SCg:byx4l5e) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/pose_selector_test.cc#L140)):*
> ```cpp
>   EXPECT_EQ(-std::numeric_limits<double>::infinity(), trailing_position.pos.s);
> }
> ```

On first read, this appears to be missing a test that the returned Odometry velocities are correct, for both the no match (infinity) and match-found cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-Kg4x-10pS__O-ivhGMS:bkyy5in)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Revised as per comments. PTAL.

---

Review status: 5 of 13 files reviewed at latest revision, 22 unresolved discussions.

---

*[drake/automotive/BUILD, line 174 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4reIbh80RSExMV2Ih:-KgF8Tb_iPtNg1WioNYK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/BUILD#L174)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `":pose_selector"`
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.cc, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4d31Hx1waC1hlj2xP:-KgF8bDlhlCovEMyN42S:bryitfp) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L15)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

I'm using it now, as of this revision (see comment below).

---

*[drake/automotive/idm_controller.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4dDK2A5EuJ2UcuOxE:-Kg5larlXWXu12HwfvFN:b20d33o) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, without reading the context, my guess is that `IdmController` should alias the geometry, and not own it.  There should be a specific reason written down if there's a reason to do it with `unique_ptr` here.
</blockquote></details>

Absolutely. (I had spotted this earlier but failed to fix on my own.)  Fixed now. 

---

*[drake/automotive/idm_controller.cc, line 40 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4dlUJMWLskVOmp6rb:-KgF9Md-MKcERqKrlyCB:b-dn443p) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L40)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should the `IdmController` output a `DrivingCommand` or just a desired acceleration?
</blockquote></details>

So, for the "spiral 0" implementation, I had intended to make `IdmController` work directly with `SimpleCar` without any intermediary translation elements (as per #5288). So, that's why it's currently outputting a `DrivingCommand`, with no steering (until I have `PurePursuit` ready, this combo will only work for straight roads). 

I've updated the doxygen header to make this clearer.

---

*[drake/automotive/idm_controller.cc, line 111 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4f3okJEFAKv7Kqu5L:-KgFKXbTXURlw7uVrxYJ:b-z2yhwb) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L111)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, have we documented anywhere that this class assumes the vehicle is traveling forward along its positive x axis?

(Ditto regarding the lead car a few lines below.)
</blockquote></details>

Fixed (see response to Jeremy's comment below).

---

*[drake/automotive/idm_controller.cc, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4ipA_5Q1w_BQNM2WU:-KgFLlWyc9XI9TcTH9qb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L115)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a comment stating that this is to avoid a singularity in IDM's math when the distance equals zero.
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4j9-45wJiRpsAZOnz:-KgFFnvVHgeBEOscUjO-:b-bqhi8c) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L129)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is the above temporary? If so, should a TODO be added that steering to follow a turning road should eventually be supported?
</blockquote></details>

I hadn't envisaged `IdmController` to do any steering (but could be convinced otherwise!). The rationale is that the steering controller likely will have to maintain state, and so it should be its own separate system. `PurePursuitController` (once implemented) would be one such controller (It's a "System Concept" in the Automotive Planner System Models doc.)

---

*[drake/automotive/idm_controller.cc, line 140 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4eK1QfHJ7DIc7kVfw:-KgFDJI3mBb9y7pO-DX-:bsn71qb) ([raw file](https://github.com/robotlocomotion/drake/blob/763eef4f7e6869b9387244272487fa0790282fe8/drake/automotive/idm_controller.cc#L140)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since `params` is not a `std::unique_ptr`, do we need the `std::move()`.
</blockquote></details>

Well, it's still a vector of `std::unique_ptr`, and so the elements of the vector are move-only. And, since `std::move` works for vectors of `std::unique_ptr`s, I think it does the trick. (See here: http://en.cppreference.com/w/cpp/utility/move)

---

*[drake/automotive/idm_controller.h, line 64 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4uedwZ8UeEIfdIA8n:-KgFEpSgn2pog6QDFciP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/idm_controller.h#L64)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is the deprecated flavor of the override, so at a minimum should change to use the new spelling.  BTW also the default is `true` anyway, so this method is not required / not helpful.
</blockquote></details>

Done.

---

*[drake/automotive/idm_controller.h, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4vMpGfQ1YWrUioZMO:-KgFLnr_I2cnZtGdKbqa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/idm_controller.h#L72)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW private methods should use `//` not `///`.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.cc, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tsXpSdsZPMn0vG_T:-KgFF3T-x4D2NcJbLX6t:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.cc#L60)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Multi-line if-else bodies require curly braces, per cppguide.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tQ8VcEDfL4fYX08q:-KgFFK07jSX3P63-aANU:bua0evr) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L18)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `world-relative velocity vector`, yes?  Or just `velocity`, since the meaning of a `FrameVelocity` is well-documented.
</blockquote></details>

Indeed that's the case! I've reworded this sentence and have made a few changes in `IdmController` to fix how the ego and traffic velocities to reflect this fact. 

---

*[drake/automotive/pose_selector.h, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4sXOkxSPT69s5gn9w:-KgFLwDzyYeS4zD3-jDw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `traffic_poses`
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.h, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4V25wnOcKHluTJAYs:-KgFM6HoIfyR0sxKmIbf:box7yz7) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L40)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, if you change "CalcRoadPosition" to be "maliput::api::Lane::CalcRoadPosition()", I believe Doxygen will make it a link.
</blockquote></details>

Actually it's local to `PoseSelector` (and it's not a member of `api::Lane`), so I think doxygen should work?

---

*[drake/automotive/pose_selector.h, line 57 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4siSK0mbuaehdrblG:-KgFMvIoUoW2JiS408JU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/pose_selector.h#L57)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I think another TODO is to support vehicles traveling against-s.
</blockquote></details>

Done.

---

*[drake/automotive/test/idm_controller_test.cc, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4t1ULDVfDcXnOAJvP:-KgFN0P-V8G6TN77TUNr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: alpha sort.
</blockquote></details>

Done.

---

*[drake/automotive/test/idm_controller_test.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tWejjacbNAFmQkOB:-KgFNHIj2TE4573ZIcCP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L51)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should "speed_offset" be "relative_speed"?
</blockquote></details>

Done.

---

*[drake/automotive/test/idm_controller_test.cc, line 55 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tHg5-nJ7AXQ3uarX:-KgFNcQcj4ejtd_xWzrc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L55)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should we also demand that `ego_speed >= 0`?
</blockquote></details>

Done.

---

*[drake/automotive/test/idm_controller_test.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4tyJdcLOLDEJnCaHd:-KgFNsMi4PDWKqljx7wc:b-bezce2) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L68)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, where did you get these labels? In `FrameVelocity`, I see: `ωx, ωy, ωz, vx, vy, vz`.
</blockquote></details>

Sorry. I'm mixing in terminology standard dynamics textbooks use here (which is inconsistent with `FrameVelocity`'s definition). Fixed.

---

*[drake/automotive/test/idm_controller_test.cc, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4zaLX7ms0u_C4Jmy-:-KgFQoiEVyZJpVsypuIB:bfmv6dj) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L106)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should this unit test also cover the scenario where the `distance_lower_limit` is hit?
</blockquote></details>

Done. (below)

---

*[drake/automotive/test/idm_controller_test.cc, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4zDe7ZiGkYA9NawV4:-KgFT2MlBPK3cNbKpEpj:bfmv6dj) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/idm_controller_test.cc#L106)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should the scenario where there is no traffic in the ego car's lane be tested?
</blockquote></details>

Done. (below)

---

*[drake/automotive/test/pose_selector_test.cc, line 140 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-Kg4u9aurAMEbcxIrN3-:-KgFgTe_clVm6_FYPAPq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d98d27b85dcea2c4f76de5f0371f2660288733d/drake/automotive/test/pose_selector_test.cc#L140)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

On first read, this appears to be missing a test that the returned Odometry velocities are correct, for both the no match (infinity) and match-found cases.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-KgGFNi_G2f3dyB0hGe4:biz9db5)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging) 

Checkpoint.  Good progress.  It will probably be this evening before I review the final two files.

---

Reviewed 6 of 8 files at r3.
Review status: 11 of 13 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-KgGJEsVa8Uc1z6Y5PVu:b-43swek)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 5 of 8 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/test/idm_controller_test.cc, line 55 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgH30y5ZMmATCQ1h9d2:-KgH30y5ZMmATCQ1h9d3:b-3sdnbu) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L55)):*
> ```cpp
>   void SetDefaultPoses(const double ego_speed, const double s_offset = 0.,
>                        const double relative_speed = 0.,
>                        bool nonzero_y_speed = false) {
> ```

BTW, consider updating the above comment to describe parameter `nonzero_y_speed`.

---

*[drake/automotive/test/idm_controller_test.cc, line 82 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgH4c-kUvElphlSBYTT:-KgH4c-kUvElphlSBYTU:bcxs5ou) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L82)):*
> ```cpp
>     traffic_poses.set_pose(kLeadIndex, Eigen::Isometry3d(translation_lead));
>     velocity[3] += relative_speed;
>     velocity[4] = (nonzero_y_speed) ? relative_speed : 0.;
> ```

I don't understand this. Explain?

---

*[drake/automotive/test/idm_controller_test.cc, line 144 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgHVPUD4IZviJxYw2NN:-KgHVPUD4IZviJxYw2NO:b-jpj9jg) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L144)):*
> ```cpp
> 
>   // Set the ego car to be alone on the road.
>   SetDefaultPoses(10. /* ego_speed */);
> ```

I'm confused by how the above sets the ego car to be alone on the road. Does this set `s_offset = 0`, `relative_speed = 0`, and `nonzero_y_speed = false` meaning the ado vehicle is literally on top of the ego vehicle?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/test/idm_controller_test.cc, line 82 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgH4c-kUvElphlSBYTT:-KgI2hZ0KtkANKTH8Iqp:bwvlsj6) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L82)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I don't understand this. Explain?
</blockquote></details>

Maybe just add a comment saying any non-zero value here would do? What if `nonzer-y_speed == true` and `relative_speed == 0`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/test/idm_controller_test.cc, line 55 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgH30y5ZMmATCQ1h9d2:-KgKyncm7XmIcnmxHkRc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L55)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider updating the above comment to describe parameter `nonzero_y_speed`.
</blockquote></details>

Done.

---

*[drake/automotive/test/idm_controller_test.cc, line 82 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgH4c-kUvElphlSBYTT:-KgKxmixbfE2lu58OyQ-:bjo8293) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L82)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Maybe just add a comment saying any non-zero value here would do? What if `nonzer-y_speed == true` and `relative_speed == 0`?
</blockquote></details>

I've changed this part to deal more naturally with a non-zero `r`-speeds, which I hope is more self-explanatory. PTAL.  

---

*[drake/automotive/test/idm_controller_test.cc, line 144 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgHVPUD4IZviJxYw2NN:-KgKyrQTCsicSGRwYv9L:b1wk5fk) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L144)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I'm confused by how the above sets the ego car to be alone on the road. Does this set `s_offset = 0`, `relative_speed = 0`, and `nonzero_y_speed = false` meaning the ado vehicle is literally on top of the ego vehicle?
</blockquote></details>

Exactly, although the velocities do not matter. This trick works because the `PoseBundle` could include the ego car's pose (or copies), which must be filtered out. I've clarified with a comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r3, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/test/idm_controller_test.cc, line 144 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-KgHVPUD4IZviJxYw2NN:-KgL52GruLlrgc3lghTg:b-xk06xx) ([raw file](https://github.com/robotlocomotion/drake/blob/c619e5f90c67a68d04d0f243d4b42b764754cb77/drake/automotive/test/idm_controller_test.cc#L144)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Exactly, although the velocities do not matter. This trick works because the `PoseBundle` could include the ego car's pose (or copies), which must be filtered out. I've clarified with a comment. 
</blockquote></details>

Wow. Tricky! Thanks for the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-KgL5CgU-pwRhFADINvq:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-KgLRBoAlXKLjTnw99pU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
@jwnimmer-tri, I think it needs your :honk honk: blessing too?

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5618#-:-KgLSVR4YSo9hBMwn7mR:b-61umn8)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
No, the only interlock still blocking is the "curate..." label.
